### ISSUE TITLE
Dataframe merge fails with KeyError

### ISSUE THREAD

**Issue by guillemborrell** [dask/dask#7178]
**What happened**:

Dask optimization fails when trying to merge two dataframes

**What you expected to happen**:

The two dataframes are merged

**Minimal Complete Verifiable Example**:

Load a reasonably-large dataset to the cluster
```
base_uri = "http://datasets.imdbws.com"
file_list = [
    "name.basics.tsv.gz",
    "title.basics.tsv.gz",
    "title.principals.tsv.gz",
    "title.ratings.tsv.gz"
]
for file_name in file_list:
    (dd.read_csv("/".join((base_uri, file_name)),
                 sep="\t", 
                 blocksize=None, 
                 dtype={
                     'isOriginalTitle': 'object', 
                     'isAdult': 'object',
                     'startYear': 'object'
                 }
                )
     .repartition(partition_size="100MB")
     .to_csv(f"s3://imdb-datasets/{'.'.join(file_name.split('.')[:2])}.csv.gz",
             compression="gzip",
             storage_options=storage_options())
    )
```

Load the files as Dask dataframes

```
people = dd.read_csv("s3://imdb-datasets/name.basics.csv.gz/*", 
                     compression="gzip",
                     blocksize=None,
                     storage_options=storage_options())

movies = dd.read_csv("s3://imdb-datasets/title.basics.csv.gz/*",
                     compression="gzip",
                     blocksize=None,
                     dtype={'startYear': 'object', 'isAdult': 'object'},
                     storage_options=storage_options())

ratings = dd.read_csv("s3://imdb-datasets/title.ratings.csv.gz/*",
                      compression="gzip",
                      blocksize=None,
                      storage_options=storage_options())

principals = dd.read_csv("s3://imdb-datasets/title.principals.csv.gz/*", 
                         compression="gzip",
                         blocksize=None,
                         storage_options=storage_options())[["tconst", "nconst", "category", "job"]]
```

Do some merges

```
filtered_ratings = ratings[ratings.numVotes > 1000]

relevant_movies = filtered_ratings[["tconst"]]

filtered_movies = (
    movies
    [['tconst', 'primaryTitle', 'originalTitle', 'startYear', 'runtimeMinutes']]
    .merge(relevant_movies, on='tconst', how='inner')
)

category_index = principals.category.unique().reset_index()
job_index = principals.job.unique().reset_index()

filtered_crew = (
    principals
    .merge(category_index, on='category', how='left')
    .merge(job_index, on='job', how='left', suffixes=('_category', '_job'))
    .merge(relevant_movies, on='tconst', how='inner')
    [["tconst", "nconst", "index_category", "index_job"]]
) 
```

And if I try to merge `filtered_crew` with `people``

```
people.merge(filtered_crew, on="nconst", how="inner").head()
```

I get the following error

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-6-4b2a05156842> in <module>
----> 1 people.merge(filtered_crew, on="nconst", how="inner").head()

~/.local/lib/python3.8/site-packages/dask/dataframe/core.py in head(self, n, npartitions, compute)
   1034             Whether to compute the result, default is True.
   1035         """
-> 1036         return self._head(n=n, npartitions=npartitions, compute=compute, safe=True)
   1037 
   1038     def _head(self, n, npartitions, compute, safe):

~/.local/lib/python3.8/site-packages/dask/dataframe/core.py in _head(self, n, npartitions, compute, safe)
   1067 
   1068         if compute:
-> 1069             result = result.compute()
   1070         return result
   1071 

~/.local/lib/python3.8/site-packages/dask/base.py in compute(self, **kwargs)
    277         dask.base.compute
    278         """
--> 279         (result,) = compute(self, traverse=False, **kwargs)
    280         return result
    281 

~/.local/lib/python3.8/site-packages/dask/base.py in compute(*args, **kwargs)
    553     )
    554 
--> 555     dsk = collections_to_dsk(collections, optimize_graph, **kwargs)
    556     keys, postcomputes = [], []
    557     for x in collections:

~/.local/lib/python3.8/site-packages/dask/base.py in collections_to_dsk(collections, optimize_graph, optimizations, **kwargs)
    329         for opt, val in groups.items():
    330             dsk, keys = _extract_graph_and_keys(val)
--> 331             dsk = opt(dsk, keys, **kwargs)
    332 
    333             for opt in optimizations:

~/.local/lib/python3.8/site-packages/dask/dataframe/optimize.py in optimize(dsk, keys, **kwargs)
     32     if fuse_subgraphs is None:
     33         fuse_subgraphs = True
---> 34     dsk, _ = fuse(
     35         dsk,
     36         keys,

~/.local/lib/python3.8/site-packages/dask/optimization.py in fuse(dsk, keys, dependencies, ave_width, max_width, max_height, max_depth_new_edges, rename_keys, fuse_subgraphs)
    598             child = children_stack[-1]
    599             if child != parent:
--> 600                 children = reducible & deps[child]
    601                 while children:
    602                     # Depth-first search

KeyError: ('assign-6806e9502a8527bf71183e8bcc82e738', 48)
```

**Anything else we need to know?**:

I can get the same result successfully by not using the `filtered_crew` dataframe

```
relevant_people = principals.merge(relevant_movies, on="tconst", how='inner')[['nconst']].drop_duplicates()
filtered_people = people.merge(relevant_people, on="nconst", how="inner")
filtered_people.head()
```

**Environment**:

- Dask version: 2021.1.0
- Python version: 3.8.0
- Operating System: Linux
- Install method (conda, pip, source): Dask helm chart

**Comment by rubenvdg** [comment]
What is `storage_options=storage_options()` ?

**Comment by guillemborrell** [comment]
The function just returns the credentials and endpoints to read from a S3 bucket.

```
{
    "key": "AWS_ACCESS_KEY_ID",
    "secret": "AWS_SECRET_ACCESS_KEY",
    "client_kwargs": {"endpoint_url": "http://mys3endpoint"},
}
```

**Comment by rubenvdg** [comment]
Could you confirm that

```
ddf1 = (
    dd.read_csv(
        "http://datasets.imdbws.com/name.basics.tsv.gz",
         sep="\t", 
         blocksize=None, 
         dtype={'isOriginalTitle': 'object', 'isAdult': 'object', 'startYear': 'object'}
    )
    .repartition(partition_size="100MB")
)

ddf2 = (
    dd.read_csv(
        "http://datasets.imdbws.com/title.principals.tsv.gz",
         sep="\t", 
         blocksize=None, 
         dtype={'isOriginalTitle': 'object', 'isAdult': 'object', 'startYear': 'object'}
    )
    .repartition(partition_size="100MB")
)

ddf1.merge(ddf2, on="nconst", how="inner").head()
```

reproduces your problem as well (at least here it does). Please try to make your minimum verifiable example as minimal as possible.

**Comment by guillemborrell** [comment]
Yes i does! It triggers the same KeyError. Thanks a lot for the effort @rubenvdg!

**Comment by jsignell** [comment]
It seems like this might have to do with the blockwise work that @rjzamora is doing.

**Comment by rjzamora** [comment]
Thanks for raising and working out a reproducer @guillemborrell and @rubenvdg (and thanks @jsignell for the ping).

As far as I can tell, the error is from a bug in `ShuffleLayer._cull_dependencies` (at some point a change was made in graph construction but not in the dependency-culling logic).  While investigating, I did find a `Blockwise` bug, but it was unrelated to this issue.  Please do let me know if #7213 resolves this issue.

**Comment by guillemborrell** [comment]
These changes fix the issue for me. Thanks @rjzamora!

**Comment by jrbourbeau** [comment]
Closing. Thanks for the fix @rjzamora and following up here @guillemborrell

**Comment by rjzamora** [comment]
@jrbourbeau - Note that #7213 wasn't merged quite yet :)

**Comment by jrbourbeau** [comment]
Whoops, it is now : )

**Comment by S2020BD** [comment]
I pip install dask[complete] --upgrade to 2021.2.0 but still get the same error.

**Comment by jrbourbeau** [comment]
The changes in https://github.com/dask/dask/pull/7213 haven't been released yet. If you need those updates right away, you can install Dask directly from GitHub

```
python -m pip install git+https://github.com/dask/dask.git@master
```

**Comment by S2020BD** [comment]
thank you for your timely answer. When is the next release?

**Comment by jrbourbeau** [comment]
I don't have an exact date for the next release, but Dask is generally released every few weeks. So I would expect the next release within that time frame. 

You can also use `git clone` to download the source code and install locally (https://docs.dask.org/en/latest/develop.html#download-code)

**Comment by S2020BD** [comment]
Since I need it these days, I did: git clone https://dask/dask
I could not find how to proceed from there in the documentation you pointed.
I plan to do:
python setup.py install
I will first read on to check if I can get confirmation this is ok but if you know what I should do or this is the right way, please let me know

**Comment by rubenvdg** [comment]
`python setup.py install` will do. I don't understand what you mean by "I prefer to install it for all, not just for me" though.

**Comment by S2020BD** [comment]
Thank you rubenvdg, this ran. How do I know it worked? pip freeze returns 
dask==2021.2.0
dask-glm==0.2.0
dask-ml==1.8.0
dask-yarn==0.8.1
Just as before and the run above produced:
[...]
Installed c:\python\python37\lib\site-packages\dask-0+unknown-py3.7.egg
Processing dependencies for dask==0+unknown
Searching for PyYAML==5.3.1
Best match: PyYAML 5.3.1
Adding PyYAML 5.3.1 to easy-install.pth file

Using c:\python\python37\lib\site-packages
Finished processing dependencies for dask==0+unknown

Is this ok? Can I check it is ok ()
dir C:\python\python37\lib\site-packages\*dask*
 Volume in drive C is Windows
 Volume Serial Number is 6E6B-1970

 Directory of C:\python\python37\lib\site-packages

02/20/2021  10:21 AM    <DIR>          dask
02/21/2021  XX:XX XM    <DIR>          dask-0+unknown-py3.7.egg
02/20/2021  10:21 AM    <DIR>          dask-2021.2.0.dist-info
02/08/2021  12:25 PM    <DIR>          dask_glm
02/08/2021  12:25 PM    <DIR>          dask_glm-0.2.0.dist-info
02/08/2021  12:25 PM    <DIR>          dask_ml
02/08/2021  12:25 PM    <DIR>          dask_ml-1.8.0.dist-info
02/20/2021  03:38 PM    <DIR>          dask_yarn
02/20/2021  03:38 PM    <DIR>          dask_yarn-0.8.1.dist-info

**Comment by rubenvdg** [comment]
@rjzamora When I install from source (with pip) in a fresh virtual env with Python 3.8.6, I catch a different error now:

```
---------------------------------------------------------------------------
error                                     Traceback (most recent call last)
<ipython-input-3-94c5aea41cf8> in <module>
     21 )
     22 
---> 23 ddf1.merge(ddf2, on="nconst", how="inner").head()

~/Desktop/tmp/dask/dask/dataframe/core.py in head(self, n, npartitions, compute)
   1040             Whether to compute the result, default is True.
   1041         """
-> 1042         return self._head(n=n, npartitions=npartitions, compute=compute, safe=True)
   1043 
   1044     def _head(self, n, npartitions, compute, safe):

~/Desktop/tmp/dask/dask/dataframe/core.py in _head(self, n, npartitions, compute, safe)
   1073 
   1074         if compute:
-> 1075             result = result.compute()
   1076         return result
   1077 

~/Desktop/tmp/dask/dask/base.py in compute(self, **kwargs)
    280         dask.base.compute
    281         """
--> 282         (result,) = compute(self, traverse=False, **kwargs)
    283         return result
    284 

~/Desktop/tmp/dask/dask/base.py in compute(*args, **kwargs)
    562         postcomputes.append(x.__dask_postcompute__())
    563 
--> 564     results = schedule(dsk, keys, **kwargs)
    565     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    566 

~/.pyenv/versions/3.8.6/envs/tmp/lib/python3.8/site-packages/distributed/client.py in get(self, dsk, keys, workers, allow_other_workers, resources, sync, asynchronous, direct, retries, priority, fifo_timeout, actors, **kwargs)
   2653                     should_rejoin = False
   2654             try:
-> 2655                 results = self.gather(packed, asynchronous=asynchronous, direct=direct)
   2656             finally:
   2657                 for f in futures.values():

~/.pyenv/versions/3.8.6/envs/tmp/lib/python3.8/site-packages/distributed/client.py in gather(self, futures, errors, direct, asynchronous)
   1962             else:
   1963                 local_worker = None
-> 1964             return self.sync(
   1965                 self._gather,
   1966                 futures,

~/.pyenv/versions/3.8.6/envs/tmp/lib/python3.8/site-packages/distributed/client.py in sync(self, func, asynchronous, callback_timeout, *args, **kwargs)
    836             return future
    837         else:
--> 838             return sync(
    839                 self.loop, func, *args, callback_timeout=callback_timeout, **kwargs
    840             )

~/.pyenv/versions/3.8.6/envs/tmp/lib/python3.8/site-packages/distributed/utils.py in sync(loop, func, callback_timeout, *args, **kwargs)
    338     if error[0]:
    339         typ, exc, tb = error[0]
--> 340         raise exc.with_traceback(tb)
    341     else:
    342         return result[0]

~/.pyenv/versions/3.8.6/envs/tmp/lib/python3.8/site-packages/distributed/utils.py in f()
    322             if callback_timeout is not None:
    323                 future = asyncio.wait_for(future, callback_timeout)
--> 324             result[0] = yield future
    325         except Exception as exc:
    326             error[0] = sys.exc_info()

~/.pyenv/versions/3.8.6/envs/tmp/lib/python3.8/site-packages/tornado/gen.py in run(self)
    760 
    761                     try:
--> 762                         value = future.result()
    763                     except Exception:
    764                         exc_info = sys.exc_info()

~/.pyenv/versions/3.8.6/envs/tmp/lib/python3.8/site-packages/distributed/client.py in _gather(self, futures, errors, direct, local_worker)
   1827                             exc = CancelledError(key)
   1828                         else:
-> 1829                             raise exception.with_traceback(traceback)
   1830                         raise exc
   1831                     if errors == "skip":

~/Desktop/tmp/dask/dask/bytes/core.py in read_block_from_file()
    185     with copy.copy(lazy_file) as f:
    186         if off == 0 and bs is None:
--> 187             return f.read()
    188         return read_block(f, off, bs, delimiter)

~/.pyenv/versions/3.8.6/lib/python3.8/gzip.py in read()
    290             import errno
    291             raise OSError(errno.EBADF, "read() on write-only GzipFile object")
--> 292         return self._buffer.read(size)
    293 
    294     def read1(self, size=-1):

~/.pyenv/versions/3.8.6/lib/python3.8/gzip.py in read()
    485             buf = self._fp.read(io.DEFAULT_BUFFER_SIZE)
    486 
--> 487             uncompress = self._decompressor.decompress(buf, size)
    488             if self._decompressor.unconsumed_tail != b"":
    489                 self._fp.prepend(self._decompressor.unconsumed_tail)

error: Error -3 while decompressing data: invalid stored block lengths
```

To reproduce:

```
from dask.distributed import Client, LocalCluster
cluster = LocalCluster()
client = Client(cluster)

import dask.dataframe as dd

ddf1 = (
    dd.read_csv(
        "http://datasets.imdbws.com/name.basics.tsv.gz",
         sep="\t", 
         blocksize=None, 
         dtype={'isOriginalTitle': 'object', 'isAdult': 'object', 'startYear': 'object'}
    )
    .repartition(partition_size="100MB")
)

ddf2 = (
    dd.read_csv(
        "http://datasets.imdbws.com/title.principals.tsv.gz",
         sep="\t", 
         blocksize=None, 
         dtype={'isOriginalTitle': 'object', 'isAdult': 'object', 'startYear': 'object'}
    )
    .repartition(partition_size="100MB")
)

ddf1.merge(ddf2, on="nconst", how="inner").head()
```

**Comment by S2020BD** [comment]
so the fix does not work quite yet? I have python 3.7.6
do you also get dask-0+unknown-py3.7.egg when running pip install setup.py, @rubenvdg ? 

with the same code, I get:

`---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-2-077ceff9c27c> in <module>
     16          dtype={'isOriginalTitle': 'object', 'isAdult': 'object', 'startYear': 'object'}
     17     )
---> 18     .repartition(partition_size="100MB")
     19 )
     20 

c:\python\python37\lib\site-packages\dask\dataframe\core.py in repartition(self, divisions, npartitions, partition_size, freq, force)
   1216 
   1217         if partition_size is not None:
-> 1218             return repartition_size(self, partition_size)
   1219         elif npartitions is not None:
   1220             return repartition_npartitions(self, npartitions)

c:\python\python37\lib\site-packages\dask\dataframe\core.py in repartition_size(df, size)
   6095     size = int(size)
   6096 
-> 6097     mem_usages = df.map_partitions(total_mem_usage, deep=True).compute()
   6098 
   6099     # 1. split each partition that is larger than partition_size

c:\python\python37\lib\site-packages\dask\base.py in compute(self, **kwargs)
    279         dask.base.compute
    280         """
--> 281         (result,) = compute(self, traverse=False, **kwargs)
    282         return result
    283 

c:\python\python37\lib\site-packages\dask\base.py in compute(*args, **kwargs)
    561         postcomputes.append(x.__dask_postcompute__())
    562 
--> 563     results = schedule(dsk, keys, **kwargs)
    564     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    565 

c:\python\python37\lib\site-packages\distributed\client.py in get(self, dsk, keys, restrictions, loose_restrictions, resources, sync, asynchronous, direct, retries, priority, fifo_timeout, actors, **kwargs)
   2679                     should_rejoin = False
   2680             try:
-> 2681                 results = self.gather(packed, asynchronous=asynchronous, direct=direct)
   2682             finally:
   2683                 for f in futures.values():

c:\python\python37\lib\site-packages\distributed\client.py in gather(self, futures, errors, direct, asynchronous)
   1994                 direct=direct,
   1995                 local_worker=local_worker,
-> 1996                 asynchronous=asynchronous,
   1997             )
   1998 

c:\python\python37\lib\site-packages\distributed\client.py in sync(self, func, asynchronous, callback_timeout, *args, **kwargs)
    835         else:
    836             return sync(
--> 837                 self.loop, func, *args, callback_timeout=callback_timeout, **kwargs
    838             )
    839 

c:\python\python37\lib\site-packages\distributed\utils.py in sync(loop, func, callback_timeout, *args, **kwargs)
    338     if error[0]:
    339         typ, exc, tb = error[0]
--> 340         raise exc.with_traceback(tb)
    341     else:
    342         return result[0]

c:\python\python37\lib\site-packages\distributed\utils.py in f()
    322             if callback_timeout is not None:
    323                 future = asyncio.wait_for(future, callback_timeout)
--> 324             result[0] = yield future
    325         except Exception as exc:
    326             error[0] = sys.exc_info()

c:\python\python37\lib\site-packages\tornado\gen.py in run(self)
    733 
    734                     try:
--> 735                         value = future.result()
    736                     except Exception:
    737                         exc_info = sys.exc_info()

c:\python\python37\lib\site-packages\distributed\client.py in _gather(self, futures, errors, direct, local_worker)
   1853                             exc = CancelledError(key)
   1854                         else:
-> 1855                             raise exception.with_traceback(traceback)
   1856                         raise exc
   1857                     if errors == "skip":

c:\python\python37\lib\site-packages\dask\bytes\core.py in read_block_from_file()
    185     with copy.copy(lazy_file) as f:
    186         if off == 0 and bs is None:
--> 187             return f.read()
    188         return read_block(f, off, bs, delimiter)

c:\python\python37\lib\gzip.py in read()
    274             import errno
    275             raise OSError(errno.EBADF, "read() on write-only GzipFile object")
--> 276         return self._buffer.read(size)
    277 
    278     def read1(self, size=-1):

c:\python\python37\lib\gzip.py in read()
    452                 # Check the CRC and file size, and set the flag so we read
    453                 # a new member
--> 454                 self._read_eof()
    455                 self._new_member = True
    456                 self._decompressor = self._decomp_factory(

c:\python\python37\lib\gzip.py in _read_eof()
    499         if crc32 != self._crc:
    500             raise OSError("CRC check failed %s != %s" % (hex(crc32),
--> 501                                                          hex(self._crc)))
    502         elif isize != (self._stream_size & 0xffffffff):
    503             raise OSError("Incorrect length of data produced")

OSError: CRC check failed 0xd4e9a5c0 != 0x8522e109`
